### ISSUE TITLE
【質問】【第3章：p.172、173】3-7_PSPNet_training の optimizer.zero_grad() について

### ISSUE THREAD

**Issue by mitgrad2018** [YutaroOgawa/pytorch_advanced#103]
小川様

optimizer.zero_grad() について質問がございます。
下記コードの通り、p.173に optimizer.zero_grad() があるので、p.172の optimizer.zero_grad() は不要でしょうか？

<p.172>
----------------------------------------------------------
        # epochごとの訓練と検証のループ
        for phase in ['train', 'val']:
            if phase == 'train':
                net.train()  # モデルを訓練モードに
                scheduler.step()  # 最適化schedulerの更新
                **optimizer.zero_grad()**
                print('（train）')
-----------------------------------------------------------

<p.173>
-----------------------------------------------------------
                # multiple minibatchでのパラメータの更新
                if (phase == 'train') and (count == 0):
                    optimizer.step()
                    **optimizer.zero_grad()**
                    count = batch_multiplier
-----------------------------------------------------------

**Comment by YutaroOgawa** [comment]
@mitgrad2018 さま

ご質問ありがとうございます。
こちら自信がないのですが、
p.172の optimizer.zero_grad() は残しておいた方が良いです。

というのも、これを消すと、今回、multiple minibatchなので、

一番初回に、p.173、# multiple minibatchでのパラメータの更新の

optimizer.step()が、
optimizer.zero_grad()が初めて実行されるよりも先に実行されてしまいます。

ちょっと、これ（optimizer.zero_grad()が一度も走る前に、optimizer.step()が走る現象）
が怖いなと思って、今の実装になっています。

毎回、非常に丁寧にコードを追って、鋭いご質問を誠にありがとうございます。
どうぞよろしくお願い致します。

**Comment by mitgrad2018** [comment]
小川様

早速ご返信いただきありがとうございます。
小川様のご懸念理解いたしました。

例えばですが、1章 1-5_fine_tuning.ipynb のように、optimizer.step() を loss.backward() の後に持ってくるのはいかがでしょうか。そうすれば、p.173の optimizer.zero_grad() が optimizer.step() より前に走ることになり、p.172の optimizer.zero_grad() を削除できるのではないかと思っております。

**Comment by YutaroOgawa** [comment]
@mitgrad2018 さま

しっかりといろいろな手段を考えてくださりありがとうございます。

ただ、ここで、
optimizer.zero_grad() が optimizer.step() より後にあるのは、multipleminibatch分繰り返されて、
たまったときに、
optimizer.step()し、その後、optimizer.zero_grad() が走って欲しいので今の実装になっています。

提案の通りに逆にすると、たまってこのif文のなかに入ったときに、step()される前にzero_grad()されてしまう・・・

本書の構造は引用で参考にした、以下と同じ構造です。
https://medium.com/@davidlmorton/increasing-mini-batch-size-without-increasing-memory-6794e10db672


「なんか、p. 172のoptimizer.zero_grad() が突然気持ち悪いな～」という気持ちはとても同意なので、
良い実装があれば、もっと美しくしたいとは思うところです。

**Comment by mitgrad2018** [comment]
小川様

ご回答いただきありがとうございます。
確かにご指摘の通りですね。optimizer.step() を loss.backward() よりも後にもってくる場合も、結局 multipleminibatch用に if文を設けてその中に入れることになり、美しくならないですね。引用元のリンクについてもありがとうございます。
ご丁寧なご回答に感謝いたします。

**Comment by YutaroOgawa** [comment]
@mitgrad2018 さま

multipleminbatchはややこしいので、きっともっと美しくなる方法はあると思うのですが。。。
私も勉強になりました。
非常に鋭い、楽しい質問をありがとうございます！！
今後ともどうぞよろしくお願い致します。
